### PR TITLE
[Exceptions] Added New BadRequest and Forbidden Exceptions

### PR DESCRIPTION
**Pull request by HenriRabalais**
## Summary
Added new BadRequest and Forbidden Exceptions for various uses when validating requests.

I am currently using these in the Biobank module and need them to be on major so that the module can be based off of major

## BadRequest
Can be thrown when the user submitted data is invalid or does not meet certain requirements.

## Forbidden 
Can be used when an action shouldn't be taken by the user based on their permissions.

### CODE REVIEW

File: php/exceptions/Forbidden.class.inc [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * @category Errors
+ * @package  Loris
+ * @author   Loris Team <loris-dev@bic.mni.mcgill.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+
+/**
+ * The NotFound exception is thrown when a resource can not
+ * be found. It maps to a 404 error.
```
Comment (johnsaigle) [inline php/exceptions/Forbidden.class.inc]:
```suggestion
 * be loaded due to insufficient permissions. It maps to a 403 error.
```

File: php/exceptions/Forbidden.class.inc [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+ * This file contains the NotFound exception type.
```
Comment (johnsaigle) [inline php/exceptions/Forbidden.class.inc]:
```suggestion
 * This file contains the Forbidden exception type.
```

File: php/exceptions/Forbidden.class.inc [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (johnsaigle) [inline php/exceptions/Forbidden.class.inc]:
```suggestion
<?php declare(strict_types=1);
```

File: php/exceptions/Forbidden.class.inc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @category Errors
+ * @package  Loris
+ * @author   Loris Team <loris-dev@bic.mni.mcgill.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+
+/**
+ * The NotFound exception is thrown when a resource can not
```
Comment (johnsaigle) [inline php/exceptions/Forbidden.class.inc]:
```suggestion
 * The Forbidden exception is thrown when a resource can not
```

File: php/exceptions/BadRequest.class.inc [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @category Errors
+ * @package  Loris
+ * @author   Loris Team <loris-dev@bic.mni.mcgill.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+
+/**
+ * The NotFound exception is thrown when a resource can not
```
Comment (johnsaigle) [inline php/exceptions/BadRequest.class.inc]:
```suggestion
 * The BadRequest exception is thrown when the user's input is invalid or insufficient.
```

File: php/exceptions/BadRequest.class.inc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * be found. It maps to a 400 error.
```
Comment (johnsaigle) [inline php/exceptions/BadRequest.class.inc]:
```suggestion
 * It maps to a 400 error.
```

File: php/exceptions/BadRequest.class.inc [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (johnsaigle) [inline php/exceptions/BadRequest.class.inc]:
```suggestion
<?php declare(strict_types=1);
```

**Comment by HenriRabalais** [comment]
@johnsaigle thanks! fixed!

File: php/exceptions/BadRequest.class.inc [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * @category Errors
+ * @package  Loris
+ * @author   Loris Team <loris-dev@bic.mni.mcgill.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+
+/**
+ * The BadRequest exception is thrown when the user's input is invalid or
+ * insufficient.
```
[the remark this answers is not in the document]
Comment (HenriRabalais) [reply php/exceptions/BadRequest.class.inc]: I think I meant that the data was missing information that was required. Permission issues would trigger a `Forbidden` exception. Bad Request is to be used exclusively for data submission.

[the remark this answers is not in the document]
Comment (PapillonMcGill) [reply php/exceptions/BadRequest.class.inc]: Could you add some text to complement insufficient?

**Comment by HenriRabalais** [comment]
@PapillonMcGill please take another look!

**Comment by xlecours** [comment]
> It would make more sense to use middleware to map known exceptions to their HTTP counterpart and include typed exceptions that are meaningful in and of themselves. 

This seems to be in sync with what Zend propose in https://docs.zendframework.com/zend-httphandlerrunner/runner/#server-request-error-response-generator

**Comment by xlecours** [comment]
@HenriRabalais I which context are you using those exceptions?

**Comment by HenriRabalais** [comment]
@xlecours I am using a BadRequest exception when the user-provide data does not meet the required validation and I am using a Forbidden exception when the user does not having permission for the given request.

**Comment by xlecours** [comment]
> @xlecours I am using a BadRequest exception when the user-provide data does not meet the required validation and I am using a Forbidden exception when the user does not having permission for the given request.

@HenriRabalais I am asking because if you are using RequestHandlerInterface implementations then those two condition could be check early in the handle function; then return a ResponseInterface implementation. 

If you really can't return a ResponseInterface then you can always add the exception classes under your module namespace.

**Comment by driusan** [comment]
Would a `\DomainException` be applicable for the case you wanted a `BadRequest` exception for?

**Comment by HenriRabalais** [comment]
@xlecours I am using the RequestHandlerInterface implementations but the validation checks are nested quite deeply within functions and it seems unruly to sent them back up the function chain. Maybe I haven't organized this properly, but within the handle function I am trying to resolve the request and then catch any exceptions that are thrown and at that point returning a ResponseInterface depending on the type of exception thrown.

File: php/exceptions/BadRequest.class.inc [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ * deceptive request routing). It maps to a 400 error.
+ *
+ * @category Errors
+ * @package  Loris
+ * @author   Loris Team <loris-dev@bic.mni.mcgill.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+
+class BadRequest extends LorisException
```
Comment (driusan) [inline php/exceptions/BadRequest.class.inc]: So are you going to remove this class from the PR in favour of the standard PHP exception?

Comment (HenriRabalais) [reply php/exceptions/BadRequest.class.inc]: Yes

**Comment by HenriRabalais** [comment]
Will re-open when addressed, or issue a new PR with desired changes.

**Pull request closed without merging**